src/pymoca/ast.py

```diff
@@ -908,45 +926,111 @@ def __deepcopy__(self, memo):
         return new
 
     def __repr__(self):
-        return "{}(type={!r}, name={!r})".format(type(self).__name__, self.type, self.name)
+        return "{}(name={!r}, type={!r})".format(type(self).__name__, self.name, self.type)
 
     def __str__(self):
         return '{} {}, Type "{}"'.format(type(self).__name__, self.name, self.type)
 
 
-class InstanceClass(Class):
+class InstanceElement:
+    """
+    Base class for instance elements (symbols, classes, extends).
+
+    This is the "partially instantiated element" in spec 3.5 section 5.6.1.4.
+    Includes name for lookup and type for redeclares during instantiation.
+    We include the latter two items that are also in sub-classes because we
+    want to allow use of this stand-alone as a "partial instance" for memory
+    efficiency and speed.
+    """
+
+    def __init__(
+        self,
+        ast_ref: Optional[Union[Class, Symbol]] = None,
+        modification_environment: Optional[ClassModification] = None,
+        fully_instantiated: bool = False,
+        **kwargs,
+    ):
+        """ast_ref is a reference to the AST node where this instance is defined.
+        All named keyword arguments optional for backward compatibility."""
+
+        # super().__init__() is only needed if 1st in method resolution order
+        super().__init__(**kwargs)
+
+        self.ast_ref = ast_ref
+
+        if modification_environment is not None:
+            self.modification_environment = modification_environment
+        else:
+            self.modification_environment = ClassModification()
+
+        if "name" in kwargs:
+            self.name = kwargs["name"]
+        elif ast_ref is not None:
```

Comment:
Am I correct in understanding that the _new_ code will always have `ast_ref` set, and the old code will pass `name` in as a keyword? If so, I'm rather leaning towards having an environment variable over your `use_find_name`, such that we can add assertions based on the path taken. This comment might not make sense if the "new" method still relies on the old behavior as well (can't tell yet at this stage of review).

Reply:
Anther question that will hopefully be more clear as you look at `tree.py`.